Documentation/Plugins.md

```diff
@@ -0,0 +1,339 @@
+
+A plugin declares which extension point it implements by defining the plugin's _capability_.  This determines the entry point through which SwiftPM will call it, and determines which actions the plugin can perform.
+
+Plugins have access to a representation of the package model, and plugins that define custom commands can also invoke services provided by SwiftPM to build and test products and targets defined in the package to which the plugin is applied.
+
+Every plugin runs in its own process in a sandbox that prevents network access as well as attempts to write to the file system.  Custom command plugins that need to modify the package source code can specify this requirement, and if the user approves, will have write access to the package directory.  Build tool plugins cannot modify the package source code.
+
+## Using a Package Plugin
+
+A package plugin is available in the package that defines it, as well as in any other package that has a direct dependency on the package that defines it.
+
+To get access to a plugin defined in another package, add a package dependency on the package that defines the plugin.  This will let the package access any build tool plugins and command plugins from the dependency.
+
+### Making use of a build tool plugin
+
+To make use of a build tool plugin, list its name in each target to which it should apply:
+
+```swift
+// swift-tools-version: 5.6
+import PackageDescription
+
+let package = Package(
+    name: "my-plugin-example",
+    dependencies: [
+        .package(url: "https://github.com/example/my-plugin-package.git", from: "1.0"),
+    ],
+    targets: [
+        .executableTarget(
+            name: "MyExample",
+            plugins: [
+                .plugin(name: "MyBuildToolPlugin", package: "my-plugin-package"),
+            ]
+        )
+    ]
+)
+```
+
+This will cause SwiftPM to call the plugin, passing it a simplified version of the package model for the target to which it is being applied.  Any build commands defined by the plugin will be incorporated into the build graph and will run at the appropriate time during the build.
+
+### Making use of a command plugin
+
+Unlike build tool plugins, which are invoked as needed when SwiftPM constructs the build task graph, command plugins are only invoked directly by the user.  This is done through the `swift` `package` command line interface:
+
+```shell
+❯ swift package my-plugin --my-flag my-parameter
+```
+
+Any command line arguments that appear after the invocation verb defined by the plugin are passed unmodified to the plugin — in this case, `--my-flag` and `my-parameter`.  This is commonly used in order to narrow down the application of a command to one or more targets, through the convention of one or more occurrences of a `--target` option with the name of the target(s).
+
+To list the plugins that are available within the context of a package, use the `--list` option of the `plugin` subcommand:
+
+```shell
+❯ swift package plugin --list
```

Comment:
not really feedback to this doc, but imo this would be nicer if it was `swift package plugin list`